File: Data Structures & Algorithms/search-for-word-ii/submission-5.cpp
struct TreeNode {
    TreeNode* children[26] = {nullptr};
    string* wordPtr = nullptr;
    int count = 0;
};

class Solution {
private:
    TreeNode* root;

    void addWord(string& word) {
        TreeNode* curr = root;

        for (char c : word) {
            if (!curr->children[c - 'a'])
                curr->children[c - 'a'] = new TreeNode();
            curr->count++;
            curr = curr->children[c - 'a'];
        }
        curr->wordPtr = &word;
    }

    void removeWord(string& word) {
        TreeNode* curr = root;

        for (char c : word) {
            curr->count--;
            curr = curr->children[c - 'a'];
        }
        curr->wordPtr = nullptr;
    }

    int dirs[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    vector<string> ans;

    void findHelper(TreeNode* node, vector<vector<char>>& board, int row, int col) {
        if (row < 0 || row >= board.size() || col < 0 || col >= board[0].size() || board[row][col] == '#')
            return;

        if (node->count <= 0)
            return;

        TreeNode* nextNode = node->children[board[row][col] - 'a'];
        if (!nextNode)
            return;
        node = nextNode;

        if (node->wordPtr) {
            ans.push_back(*node->wordPtr);
            removeWord(*node->wordPtr);
        }

        char aux = board[row][col];
        board[row][col] = '#';

        for (auto d : dirs) {
            findHelper(node, board, row + d[0], col + d[1]);
        }

        board[row][col] = aux;
    }

public: 
    Solution() {
        root = new TreeNode();
    }

    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        for (int i = 0; i < words.size(); ++i)
            addWord(words[i]);

        for (int i = 0; i < board.size(); ++i)
            for (int j = 0; j < board[0].size(); ++j)
                findHelper(root, board, i, j);

        return ans;
    }
};

File: Data Structures & Algorithms/majority-element-ii/submission-5.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        if len(nums) < 2:
            return [nums[0]]
        c1 = [nums[0], 0]
        c2 = [nums[1], 0]
        for i in range(len(nums)):
            if nums[i] == c1[0]:
                c1[1] += 1
                continue
            elif nums[i] == c2[0]:
                c2[1] += 1
                continue

            if c1[1] > 0:
                c1[1] -= 1
            elif c2[1] > 0:
                c1[0] = nums[i]

            if c2[1] > 0:
                c2[1] -= 1
            else:
                c2[0] = nums[i]
        
        c1[1] = 0
        c2[1] = 0
        for n in nums:
            if n == c1[0]:
                c1[1] += 1
            elif n == c2[0]:
                c2[1] += 1

        ans = []
        if c1[1] > len(nums) // 3:
            ans.append(c1[0])
        if c2[1] > len(nums) // 3:
            ans.append(c2[0])
        return ans
        